### ISSUE TITLE
Eviction manager: When stats calls failed, we evict the pod that we couldn't get stats for.

### ISSUE THREAD

**Issue by kannon92** [kubernetes/kubernetes#123591]
### What happened?

As I was investigating the eviction tests, I discovered an interesting problem. We are seeing a wide range of flakes where the one pod that should not be evicted gets evicted in the eviction test.  

Eviction tests work by assigning one pod that it should never be evicted and then it ranks these pods and evicts the other pods. If there is a stats API failure, we evict the pod that we couldn't get stats for.

This seems to be by design (cc @tallclair)

### What did you expect to happen?

We should maybe consider a more deterministic approach if stats fails.

### How can we reproduce it (as minimally and precisely as possible)?

Try running the eviction tests on the PR.

### Anything else we need to know?

Code that causes this is here:

https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/eviction/helpers.go#L747

### Kubernetes version

all supported

### Cloud provider

na

### OS version

na

### Install tools

na

### Container runtime (CRI) and version (if applicable)

crio and containerd

**Comment by kannon92** [comment]
/triage accepted

This seems to be one of the causes for eviction test flakiness.

I've seen in mostly in disk stats.

**Comment by kannon92** [comment]
Some ideas could be to use QoS in case of stats failure. We could say that BestEffort gets evicted before Burstable and Burstable before Guaranteed. This could help in the case of stats failure so we can control the eviction a bit better.

An interesting thing about this is this doesn't help disk based eviction as ephemeral-storage is never considered guaranteed (guaraneteed is only for cpu/memory).

**Comment by kannon92** [comment]
cc @pacoxu

**Comment by neolit123** [comment]
/sig node
?

**Comment by mikebrow** [comment]
Yeah overloading the compare prioritization with error checking appears on face to be problematic.  I'd try to split that out into two separate funcs one for get stats another for the compare.  The inability to get stats for pod a may be due to a block of the resource by pod b that will be ignored because of the failure it caused?

**Comment by kannon92** [comment]
Brought this up to sig-node.

We should look into ways to populate cache more aggressively. Qos could be an option but it is worth investigating why stats gathering fails.

**Comment by AnishShah** [comment]
I compared failed and successful runs of local storage eviction test:

**Successful Run**

During the successful run, after the culprit pod is evicted and cleaned up, the eviction manager sees that the resource is reclaimed and does not meet the threshold for eviction anymore. Hence, it does not evict any innocent pod.

![image](https://github.com/user-attachments/assets/03905d1d-d153-4e05-8743-4f4d3cc53009)


**Failed Run**

During the failed run, after the culprit pod is evicted and cleaned up, the eviction manager does not see that the resource is reclaimed and  still meets the threshold for eviction. Hence, it evicts the innocent pod next.

![image](https://github.com/user-attachments/assets/ae15a911-5bd3-4e81-80ba-fb3e7afb1f51)

This feels like a race condition between eviction manager and how aggressively cadvisor populates its cache.
As @kannon92 suggested in https://github.com/kubernetes/kubernetes/issues/123591#issuecomment-1995016731, populating cache more aggressively should help here. But it might not totally fix this race condition.

**Comment by AnishShah** [comment]
cadvisor config to sync its memory cache - https://github.com/kubernetes/kubernetes/blob/385fd21d92d1e598f9ca6b68ccd1b6c5b2797f3e/pkg/kubelet/cadvisor/cadvisor_linux.go#L64

**Comment by natasha41575** [comment]
/assign @esotsal 
to see if this is still an issue today. I don't see any updates on this issue, so I suspect that it still occurs. 
/triage accepted

**Comment by esotsal** [comment]
Yes , is still an issue today resulting on Flaky tests, [39 failures (3 in last day) out of 111386 builds from 9/9/2025, 2:00:04 AM to 9/19/2025, 8:52:06 AM](https://storage.googleapis.com/k8s-triage/index.html?sig=node&text=eviction_test#9280c5ea787e4696657e)

Checking [logs](https://storage.googleapis.com/kubernetes-ci-logs/logs/ci-kubernetes-node-kubelet-containerd-eviction/1965640811999662080/artifacts/e2-standard-2-cos-beta-125-19216-0-38-8d1dda9b/containerd.log) from a most [recent failure,](https://prow.k8s.io/view/gs/kubernetes-ci-logs/logs/ci-kubernetes-node-kubelet-containerd-eviction/1965640811999662080)  i agree with @AnishShah comments shared.

**Comment by ylink-lfs** [comment]
I reproduced this behavior against the current `master` branch. Before working on a fix, I'd like maintainer input on the expected behavior and scope.

The manager-level test has two active pods under nodefs pressure (`750Mi` available against a `1Gi` threshold). The stats summary succeeds but includes only the lower-priority pod, which is also using more disk. The pod without stats has higher priority, but the eviction manager ranks and selects it first. When I add its stats, the measured high-usage pod is selected instead. Reversing the active-pod input order does not change the result: the pod without stats is still selected. The regression case and both controls pass with `go test -race`.

The selection happens as follows. After [`summaryProvider.Get`](https://github.com/kubernetes/kubernetes/blob/bbedc834eb4e2509cb913a5e0dcf5236fab6fac9/pkg/kubelet/eviction/eviction_manager.go#L301-L319) returns a successful summary, `makeSignalObservations` builds the pod-stats lookup from `summary.Pods`. [`cachedStatsFunc`](https://github.com/kubernetes/kubernetes/blob/bbedc834eb4e2509cb913a5e0dcf5236fab6fac9/pkg/kubelet/eviction/helpers.go#L610-L620) returns `found == false` for an active pod whose UID is missing from the summary. For filesystem pressure, [`rankDiskPressureFunc`](https://github.com/kubernetes/kubernetes/blob/bbedc834eb4e2509cb913a5e0dcf5236fab6fac9/pkg/kubelet/eviction/helpers.go#L751-L833) evaluates `exceedDiskRequests` before priority. When it compares a pod without stats with a measured pod, the missing-entry branch returns a non-zero result in favor of the pod without stats. The sorter therefore does not reach the priority comparison. [`synchronize`](https://github.com/kubernetes/kubernetes/blob/bbedc834eb4e2509cb913a5e0dcf5236fab6fac9/pkg/kubelet/eviction/eviction_manager.go#L402-L450) then tries the first sorted candidate. This reproduces how omission of a whole pod entry affects selection, but it does not explain why the entry was omitted.

If pods with measurements should be preferred, I propose making stats availability the first comparison for filesystem-pressure ranking. Candidates with stats would sort ahead of candidates without them, followed by the existing request, priority, and usage comparisons. If every candidate lacks stats, availability ties and the existing priority behavior still applies. The change would add no state, caching, retries, QoS fallback, or new observability. Manager-level tests would cover the reproduced case, restored stats, reversed input, and the case where all candidates lack stats. The cAdvisor or summary-freshness question would remain a separate producer-side investigation.

I'm unsure about the intended behavior and where the fix belongs. Could I get some hint?

- When a successful summary omits stats for some active pods, should eviction prefer measured candidates, keep the current missing-first behavior, or skip eviction for that cycle?
- Should this be fixed in eviction-manager ranking, in telemetry production or caching, or in both places?
- Should the same contract apply to filesystem, memory, and PID pressure, or should the first change cover only the reproduced filesystem path?
